pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrTokenManager.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.antlr;
+
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.Lexer;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+
+import net.sourceforge.pmd.cpd.token.GenericAntlrToken;
+
+import com.beust.jcommander.internal.Nullable;
+
+import net.sourceforge.pmd.lang.TokenManager;
+
+/**
+ * Generic token manager implementation for all Antlr lexers.
+ */
+public class AntlrTokenManager implements TokenManager {
+    private final Lexer lexer;
+    private String fileName;
+    private final String commentToken;
+    private GenericAntlrToken previousComment;
+
+    /**
+     * Constructor
+     *
+     * @param lexer The lexer
+     * @param fileName The file name
+     * @param commentToken The list of all comment tokens on the grammar
+     */
+    public AntlrTokenManager(final Lexer lexer, final String fileName, @Nullable final String commentToken) {
+        this.lexer = lexer;
+        this.fileName = fileName;
+        this.commentToken = commentToken;
+        resetListeners();
+    }
+
+    @Override
+    public Object getNextToken() {
+        final Token token = lexer.nextToken();
+        if (isCommentToken(token.getText())) {
+            previousComment = new GenericAntlrToken(token, previousComment);
+        }
+
+        return new GenericAntlrToken(token, previousComment);
+    }
+
+    @Override
+    public void setFileName(String fileName) {
+        this.fileName = fileName;
+    }
+
+    public String getFileName() {
+        return fileName;
+    }
+
+    private void resetListeners() {
+        lexer.removeErrorListeners();
+        lexer.addErrorListener(new ErrorHandler());
+    }
+
+    private boolean isCommentToken(final String text) {
+        return commentToken != null && text != null && text.startsWith(commentToken);
```

Comment:
this is not gonna cut it. comments can have multiple forms (ie: single-line vs multi-line)

On antlr, the common approach is to use a channel for comments (normally HIDDEN), so you could just identify comments by which channel they come from.

An alternative would be to not use the raw lexer, but wrap it on `BufferedTokenStream`, which allows for [`getHiddenTokensToLeft()`](https://www.antlr.org/api/Java/org/antlr/v4/runtime/BufferedTokenStream.html#getHiddenTokensToLeft-int-).

This would be equivalent to what JavaCC's `Token` class assumes: comments are special tokens, not considered in the AST